### ISSUE TITLE
remember controller mappings

### ISSUE THREAD

**Issue by jordandev** [Provenance-Emu/Provenance#433]
Is it possible to have a way to remember controller player mappings?

Every time I open Provenance on my AppleTV I need to first go to the controller mappings and either remove Player 2 from the MFi controller or assign it to Player 1. It never seems to remember anything and is usually the opposite of what I want.

**Comment by jasarien** [comment]
There's probably a solution to this that I'm unaware of, but the main issue is that controllers in iOS/tvOS don't seem to have persistent unique identifiers, so there's nothing to 'remember'. The only piece of info I could get that identifies a controller was its generic name, but if you have two controllers of the same brand and model, then they will appear as the same name, so that doesn't solve the issue either.

Interested to hear if anyone has solutions for this issue.

**Comment by atyoung** [comment]
@jasarien Yeah it's a little wonky on that... You could have the user set a preferred order (with some sane inbuilt default) based on controller profile, and then loop through the connected controllers assigning the player index automatically, based on device type. Considering all the MFI controllers have indicator lights as part of the spec it shouldn't present too much difficulty for players.

**Comment by sevdestruct** [comment]
I believe this was also mentioned as part of #276, and considering the above mention of MFi controllers showing player number, I think this is moot outside of the already flawed icade connections that can't do this anyway..

**Comment by jasarien** [comment]
I think I may have stumbled across a way to remember specific controllers (at least MFi controllers, anyway).

There seems to be an undocumented API `deviceHash` on `GCController` which appears to be a unique ID for a given device.

I haven’t tested it in every case I can think of though, so there is a possibility that maybe the device has is dependant on some other info, and might change if its repaired, etc. So it might not be a 100% fool-proof solution, but it’s definitely something to think about.

**Comment by JoeMatt** [comment]
@claude might not be too hard to extend our controller selection UIs and controller manager to save a "preferred player" selection like "always", "preferred" "auto" where auto is what we have now basically, preferred is the case where you prefer player2 but only 1 controller is connected so you map to 1, but if there are two players, you priority map to 2 (which begs the question about conflicts of priority, maybe don't overthink that).

fire off a work ticket for this if it seems pretty easy

**Comment by JoeMatt** [comment]
> ## 🤔 Working on Request
> 
> ### Progress
> 
> * [x]  🚀 Agent started[ ]  📖 Reading context...[ ]  🔨 Working...[ ]  📤 Done
> 
> _Started: 2026-03-07 04:42 UTC_

@claude did you get stuck making new tickets or tagging existing tickets with this work?

we have a few epics and sub tasks now related to contorller assignment improvements function and UX wise, make sure they're all created and this ticket is update to reflect those relationships

**Comment by JoeMatt** [comment]
@claude what's the status here?
we merged a bunch of shit, haven't tested myself but are we done or have more to do other than test?
